File: app/src/main/java/com/psliusar/layers/sample/MainActivity.java
package com.psliusar.layers.sample;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.ViewGroup;

import com.psliusar.layers.LayersActivity;
import com.psliusar.layers.Transition;
import com.psliusar.layers.sample.screen.track.TracksLayer;
import com.psliusar.layers.sample.screen.child.ChildrenContainerLayer;
import com.psliusar.layers.sample.screen.dialog.DialogsLayer;
import com.psliusar.layers.sample.screen.home.HomeLayer;
import com.psliusar.layers.sample.screen.listener.ListenerLayer;
import com.psliusar.layers.sample.screen.save.SaveLayer;
import com.psliusar.layers.sample.screen.stack.StackLayer;

public class MainActivity extends LayersActivity {

    @Override
    protected void onCreate(Bundle state) {
        super.onCreate(state);
        setContentView(R.layout.activity_main);
        if (state == null) {
            getLayers().add(HomeLayer.class).setName("Home").commit();
        }
    }

    @NonNull
    @Override
    public ViewGroup getDefaultContainer() {
        return getView(R.id.container);
    }

    public void addToStack(final CharSequence title, final int level, boolean opaque) {
        getLayers().add(StackLayer.class)
                .prepareLayer(new Transition.OnLayerTransition<StackLayer>() {
                    @Override
                    public void onBeforeTransition(@NonNull StackLayer layer) {
                        layer.setParameters(title, level);
                    }
                })
                .setName("Stack" + level)
                .setOpaque(opaque)
                .setInAnimation(R.anim.lower_out, R.anim.upper_in)
                .setOutAnimation(R.anim.upper_out, R.anim.lower_in)
                .commit();
    }

    public void showChildrenLayers() {
        getLayers().add(ChildrenContainerLayer.class)
                .setName("Children")
                .commit();
    }

    public void showDialogLayers() {
        getLayers().add(DialogsLayer.class)
                .setName("Dialogs")
                .commit();
    }

    public void showActivityListener() {
        getLayers().add(ListenerLayer.class)
                .setName("Listener")
                .commit();
    }

    public void showSaveState() {
        getLayers().add(SaveLayer.class)
                .prepareLayer(new Transition.OnLayerTransition<SaveLayer>() {
                    @Override
                    public void onBeforeTransition(@NonNull SaveLayer layer) {
                        layer.setParameters("First", "Second", "Third");
                    }
                })
                .setName("Save")
                .commit();
    }

    public void showTracks() {
        getLayers().add(TracksLayer.class)
                .setName("Tracks")
                .commit();
    }
}

File: app/src/main/java/com/psliusar/layers/sample/MainActivity.kt
package com.psliusar.layers.sample

import android.content.Intent
import android.os.Bundle
import android.view.ViewGroup
import androidx.appcompat.widget.Toolbar
import com.psliusar.layers.Layer
import com.psliusar.layers.Layers
import com.psliusar.layers.LayersActivity
import com.psliusar.layers.Transition
import com.psliusar.layers.sample.screen.cards.CardsLayer
import com.psliusar.layers.sample.screen.child.ChildrenContainerLayer
import com.psliusar.layers.sample.screen.dialog.DialogsLayer
import com.psliusar.layers.sample.screen.home.HomeLayer
import com.psliusar.layers.sample.screen.listener.ListenerLayer
import com.psliusar.layers.sample.screen.save.SaveLayer
import com.psliusar.layers.sample.screen.stack.StackLayer

class MainActivity : LayersActivity(R.layout.activity_main) {

    override val defaultContainer: ViewGroup
        get() = getView(R.id.container)

    override fun onCreate(state: Bundle?) {
        super.onCreate(state)
        val toolbar = getView<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        toolbar.setNavigationOnClickListener {
            onBackPressed()
        }

        if (state == null) {
            layers.add<HomeLayer> {
                name = resources.getString(R.string.app_name)
            }
        }

        layers.addLayersEventListener(object : Layers.LayersEventListener {
            override fun onTransitionStarted(layers: Layers, transition: Transition<*>) {

            }

            override fun onTransitionFinished(layers: Layers, transition: Transition<*>) {

            }

            override fun onStackChanged(layers: Layers) {
                updateToolbar()
            }
        })
        updateToolbar()
    }

    fun addToStack(title: CharSequence, level: Int, opaque: Boolean) {
        layers.add<StackLayer> {
            withLayer { setParameters(title, level) }
            name = "Layer $level in Stack"
            this.opaque = opaque
            if (opaque) {
                setInAnimation(R.anim.lower_out, R.anim.upper_in)
                setOutAnimation(R.anim.upper_out, R.anim.lower_in)
            } else {
                setInAnimation(0, R.anim.upper_in)
                setOutAnimation(R.anim.upper_out, 0)
            }
        }
    }

    fun replaceInStack(title: CharSequence, level: Int, opaque: Boolean) {
        layers.replace<StackLayer> {
            withLayer { setParameters(title, level) }
            name = "Layer $level in Stack"
            this.opaque = opaque
            setInAnimation(R.anim.lower_out, R.anim.upper_in)
            setOutAnimation(R.anim.upper_out, R.anim.lower_in)
        }
    }

    fun showCards() {
        layers.add<CardsLayer> {
            name = "Cards Demo"
        }
    }

    fun showChildrenLayers() {
        layers.add<ChildrenContainerLayer> {
            name = "Children Layers Demo"
        }
    }

    fun showDialogLayers() {
        layers.add<DialogsLayer> {
            name = "Dialogs Layers Demo"
        }
    }

    fun showActivityListener() {
        layers.add<ListenerLayer> {
            name = "Activity Lifecycle Demo"
        }
    }

    fun showSaveState() {
        layers.add<SaveLayer> {
            withLayer { setParameters("First", "Second", "Third") }
            name = "State Saving Demo"
        }
    }

    fun showFragment() {
        startActivity(Intent(this, LayerFragmentActivity::class.java))
    }

    private fun updateToolbar() {
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(layers.stackSize > 1)
            title = layers.peek<Layer>()?.name
        }
    }
}
